vignettes/contribution.Rmd

```diff
@@ -0,0 +1,371 @@
+
+This file includes some utility function/objects.
+
+## src
+
+src folders includes all the source code written in C++.
+Files starting with `test-` are tests.
+
+### covariance.h
+
+This file includes the implementations of covariance structures with suitable parameterization,
+and the calculation of lower triangular part of Cholesky factorization of the covariance matrix.
+
+For more about the implementations and why Cholesky factorization is needed, visit [mmrm details vignette](algorithm.html#covariance-matrix-model).
+
+For more about covariance structures, visit [covariance structures vignette](covariance.html).
+
+### init.h
+
+This file includes the initialization of R and C++ communication.
+
+### mmrm.cpp
+
+This file is the core part where the likelihood is calculated.
+The objective function reads input from R and do the calculations inside.
+
+### utils.h
+
+This files includes utilities that used internally.
+
+### tmb.cpp and tmb_includes.h
+
+These two files includes the TMB module.
+
+### test- files
+
+All files starting with `test-` are unit tests for the C++ codes.
+
+## tests
+
+This folder includes all the unit tests that is done in R side, using `testthat`.
+
+## vignettes
+
+This folder includes vignettes that helps user understand more about the usage, detail, etc.
+
+## other files
+
+There are other files that can be helpful in package development.
+
+### .gitignore
+
+This file will specify untracked files/file patterns that git should ignore.
+
+### .lintr
+
+This file serves as configuration for `lintr` to do the static code analysis.
+Please install `lintr` package in your developing system to enable the analysis.
+It will provide you information about error, style issues, etc.
+
+For more, visit [lintr documentation](https://lintr.r-lib.org/).
+
+### .pre-commit-config.yaml
+
+This file includes the configurations that we use for `pre-commit`.
+`pre-commit` is a tool that help us identify simple issues before we submit our code.
+For more, visit [pre-commit documentation](https://pre-commit.com/).
+
+
+### _pkgdown.yml
+
+This file is the configuration file for `pkgdown`.
+`pkgdown` is used to build the package website easily.
+
+For more, visit [pkgdown documentation](https://pkgdown.r-lib.org/).
+
+### staged_dependencies.yaml
+
+This file is the configuration file for `staged.dependencies`.
+`staged.dependencies` is used to ensure the dependencies is correct when you have multiple repositories developing in the same time.
+
+For more, visit [staged.dependencies documentation](https://openpharma.github.io/staged.dependencies/).
+
+# Package conventions
+
+Before contributing to this package, there is some conventions to follow
+
+## Programming conventions
+
+1. function names should be explicit and clear, separated by underscore.
+2. functions starting with `h_` are supporting functions and most of them should not be exported.
+3. functions should be well documented using roxygen2 even if it is not exported.
```

Comment:
```suggestion
1. Function names should be explicit and clear, separated by underscore (`snake_case`).
2. Functions starting with `h_` are supporting functions and they should not be exported.
3. Functions should be well documented using `roxygen2` even when they are not exported.
```

Reply:
low level function `h_mmrm_tmb` is exported  (and previously `h_free_cores` and `h_mmrm_tmb_control`). Consier renaming `h_mmrm_tmb` function?

Reply:
low level function `h_mmrm_tmb` is exported  (and previously `h_free_cores` and `h_mmrm_tmb_control`). Consier renaming `h_mmrm_tmb` function and remove 'Helper Functions' section in documentation page?

Reply:
yeah currently it is like that but we will rename that (in a separate PR/issue, discussion is ongoing)